runtime/src/axum/mod.rs

```diff
@@ -0,0 +1,529 @@
+
+        Self {
+            router: Mutex::new(None),
+            logs_rx: Mutex::new(Some(rx)),
+            logs_tx: Mutex::new(tx),
+            kill_tx: Mutex::new(None),
+        }
+    }
+}
+
+impl Default for AxumWasm {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+impl Runtime for AxumWasm {
+    async fn load(
+        &self,
+        request: tonic::Request<LoadRequest>,
+    ) -> Result<tonic::Response<LoadResponse>, Status> {
+        let wasm_path = request.into_inner().path;
+        trace!(wasm_path, "loading");
+
+        let router = RouterBuilder::new()
+            .map_err(|err| Status::from_error(err.into()))?
+            .src(wasm_path)
+            .build()
+            .map_err(|err| Status::from_error(err.into()))?;
+
+        *self.router.lock().unwrap() = Some(router);
+
+        let message = LoadResponse { success: true };
+
+        Ok(tonic::Response::new(message))
+    }
+
+    async fn start(
+        &self,
+        request: tonic::Request<StartRequest>,
+    ) -> Result<tonic::Response<StartResponse>, Status> {
+        let StartRequest {
+            deployment_id,
+            port,
+            ..
+        } = request.into_inner();
+
+        let address = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), port as u16);
+
+        let logs_tx = self.logs_tx.lock().unwrap().clone();
+
+        let (kill_tx, kill_rx) = tokio::sync::oneshot::channel();
+
+        *self.kill_tx.lock().unwrap() = Some(kill_tx);
+
+        let router = self
+            .router
+            .lock()
+            .unwrap()
+            .take()
+            .context("tried to start a service that was not loaded")
+            .map_err(|err| Status::internal(err.to_string()))?;
+
+        tokio::spawn(run_until_stopped(
+            router,
+            deployment_id,
+            address,
+            logs_tx,
+            kill_rx,
+        ));
+
+        let message = StartResponse { success: true };
+
+        Ok(tonic::Response::new(message))
+    }
+
+    type SubscribeLogsStream = ReceiverStream<Result<runtime::LogItem, Status>>;
+
+    async fn subscribe_logs(
+        &self,
+        _request: tonic::Request<SubscribeLogsRequest>,
+    ) -> Result<tonic::Response<Self::SubscribeLogsStream>, Status> {
+        let logs_rx = self.logs_rx.lock().unwrap().deref_mut().take();
+
+        if let Some(logs_rx) = logs_rx {
+            Ok(tonic::Response::new(ReceiverStream::new(logs_rx)))
+        } else {
+            Err(Status::internal("logs have already been subscribed to"))
+        }
+    }
+
+    async fn stop(
+        &self,
+        request: tonic::Request<StopRequest>,
+    ) -> Result<tonic::Response<StopResponse>, Status> {
+        let request = request.into_inner();
+
+        let service_name = ServiceName::from_str(request.service_name.as_str())
+            .map_err(|err| Status::from_error(Box::new(err)))?;
+
+        let kill_tx = self.kill_tx.lock().unwrap().deref_mut().take();
```

Comment:
Does the kill channel need to be behind a mutex? Why not just set it on `.new()`?

Reply:
Good question :smile: